File: 2025/src/nim/dayThree.nim
# Copyright (C) 2025 Rémy Cases
# See LICENSE file for extended copyright information.
# This file is part of adventOfCode project from https://github.com/remyCases/adventOfCode.

import os

proc computeJoltageFromTwoBanks(batteries: seq[int64], sum_joltage: var int64) =
    var joltage: array[2, int64]
    var indeces: array[2, int64]

    for i in indeces[0]..<(batteries.len()-1):
        if batteries[i] > joltage[0]:
            joltage[0] = batteries[i]
            indeces[1] = i+1

    for i in indeces[1]..<batteries.len():
        if batteries[i] > joltage[1]:
            joltage[1] = batteries[i]

    sum_joltage += (joltage[0] * 10 + joltage[1])

proc computeJoltageFromTwelveBanks(batteries: seq[int64], sum_joltage: var int64) =
    const nBatteries = 12
    var joltage: array[nBatteries, int64]
    var indeces: array[nBatteries, int64]
    var val_joltage: int64 = 0
    
    for i in 0..<nBatteries:
        for j in indeces[i]..<(batteries.len()-nBatteries+1+i):
            if batteries[j] > joltage[i]:
                joltage[i] = batteries[j]
                if i != nBatteries-1: indeces[i+1] = j+1

        val_joltage = val_joltage * 10 + joltage[i]
    
    sum_joltage += val_joltage

proc parseBatteries(line: string, batteries: var seq[int64]): bool =
    for l in line:
        batteries.add(int64(l) - int64('0'))
    return true

proc readFileAndCompute(filename: string, option = '1') =
    var sum_joltage: int64 = 0
    var batteries: seq[int64] = newSeq[int64]()

    let compute = case option:
    of '1':
        computeJoltageFromTwoBanks
    of '2':
        computeJoltageFromTwelveBanks
    else:
        echo "Invalid part"
        return

    for line in filename.lines:
        batteries = @[]
        if not parseBatteries(line, batteries):
            echo "Error parsing ", line
            return
        compute(batteries, sum_joltage)

    echo "TOTAL JOLTAGE: ", sum_joltage

proc main*(part: char) =
    const file = "2025/data/input_day_three"
    let filename = getCurrentDir() / file
    readFileAndCompute(filename, part)

when isMainModule:
    main('1')

File: 2025/src/py/day_three.py
# Copyright (C) 2025 Rémy Cases
# See LICENSE file for extended copyright information.
# This file is part of adventOfCode project from https://github.com/remyCases/adventOfCode.

import ast
from pathlib import Path
from typing import List

NUMBER_BATTERIES = 12

def computeJoltageFromTwoBanks(batteries: List[int]) -> int:
    joltage: List[int] = [0, 0]
    indeces: List[int] = [0, 0]

    for i in range(indeces[0], (len(batteries)-1)):
        if batteries[i] > joltage[0]:
            joltage[0] = batteries[i]
            indeces[1] = i+1

    for i in range(indeces[1], len(batteries)):
        if batteries[i] > joltage[1]:
            joltage[1] = batteries[i]

    return joltage[0] * 10 + joltage[1]

def computeJoltageFromTwelveBanks(batteries: List[int]) -> int:
    
    joltage: List[int] = [0]*NUMBER_BATTERIES
    indeces: List[int] = [0]*NUMBER_BATTERIES
    val_joltage: int = 0
    
    for i in range(0, NUMBER_BATTERIES):
        for j in range(indeces[i], (len(batteries)-NUMBER_BATTERIES+1+i)):
            if batteries[j] > joltage[i]:
                joltage[i] = batteries[j]
                if i != NUMBER_BATTERIES-1:
                    indeces[i+1] = j+1

        val_joltage = val_joltage * 10 + joltage[i]
    
    return val_joltage

def parseBatteries(line: str) -> List[int]:
    batteries: List[int] = []
    for ll in line:
        if ll == '\n':
            break
        batteries.append(ast.literal_eval(ll))
    return batteries

def readFileAndCompute(filename: Path, option: int = 1) -> None:
    sum_joltage: int = 0

    with open(filename, 'r') as file:
        for line in file:
            batteries = parseBatteries(line)
            if option == 1:
                sum_joltage += computeJoltageFromTwoBanks(batteries)
            elif option == 2:
                sum_joltage += computeJoltageFromTwelveBanks(batteries)
            else:
                print("Invalid part")
                return

    print(f"TOTAL JOLTAGE: {sum_joltage}")

def main(part: int) -> None:
    file = Path("2025/data/input_day_three")
    filename = Path.cwd() / file
    readFileAndCompute(filename, part)
